commands.py

```diff
@@ -0,0 +1,222 @@
+    """
+    An asynchronous WindowCommand. Override the `execute` method instead
+    of the `run` method.
+
+    This is the exclusive task variant. Meaning, if you run this command, and
+    then run it again, the first invocation (if it's still running) will be
+    cancelled. Only one `execute` call is running at all times.
+    """
+    def __init__(self, window: sublime.Window):
+        super().__init__(window)
+        DispatchMixin.__init__(self)
+        self.__task: Optional[ref[Task]] = None
+
+    def __del__(self) -> None:
+        self.__cancel()
+        super().__del__()
+
+    def __cancel(self) -> None:
+        if self.__task:
+            if task := self.__task():
+                call_soon_threadsafe(task.cancel)
+            self.__task = None
+
+    def __store_task(self, task: Task) -> None:
+        self.__task = ref(task)
+
+    def run(self, **kwargs: Any) -> None:  # type: ignore
+        self.__cancel()
+        run_coroutine(self.execute(**kwargs), self.__store_task)
+
+    async def execute(self, **kwargs: Any) -> None:
+        """
+        The asynchronous entrypoint of your command.
+        """
+        pass
+
+
+class ConcurrentWindowCommand(sublime_plugin.WindowCommand, DispatchMixin):
+    """
+    An asynchronous WindowCommand. Override the `execute` method instead
+    of the `run` method.
+
+    This is the concurrent task variant. Meaning, if you run this command, and
+    then run it again, the first invocation will keep running.
+    """
+    def __init__(self, window: sublime.Window):
+        super().__init__(window)
+        DispatchMixin.__init__(self)
+
+    def run(self, **kwargs: Any) -> None:  # type: ignore
+        self.dispatch(self.execute(**kwargs))
+
+    async def execute(self, **kwargs: Any) -> None:
+        """
+        The asynchronous entrypoint of your command.
+        """
+        pass
+
+
+class ViewCommand(sublime_plugin.TextCommand, DispatchMixin):
+    """
+    An asynchronous ViewCommand. Override the `execute` method instead
+    of the `run` method.
+
+    This is the exclusive task variant. Meaning, if you run this command, and
+    then run it again, the first invocation (if it's still running) will be
+    cancelled. Only one `execute` call is running at all times.
+    """
+    def __init__(self, view: sublime.View):
+        super().__init__(view)
+        DispatchMixin.__init__(self)
+        self.__task: Optional[ref[Task]] = None
+
+    def __del__(self) -> None:
+        self.__cancel()
+        super().__del__()
+
+    def __cancel(self) -> None:
+        if self.__task:
+            if task := self.__task():
+                call_soon_threadsafe(task.cancel)
+            self.__task = None
+
+    def __store_task(self, task: Task) -> None:
+        self.__task = ref(task)
+
+    def run(self, **kwargs: Any) -> None:  # type: ignore
+        # async view commands cannot mutate the buffer
+        kwargs.pop("edit")
```

Comment:
`edit` argument will be in `kwargs` even if passed in non-keyword way?

Reply:
It seems wrong to me according to this simple test:

```py
def ttt(**kwargs):
    print(kwargs)


ttt('test')
```

This code will crash because we're passing positional argument and `**kwargs` doesn't catch that.

Reply:
`run` and `execute` are never passed positional args

Reply:
Can you explain more? I thought that calling `view.run_command('something')` will result in ST internally calling the `run` method with a positional `edit` argument. Does it work differently in this case?

Reply:
You’re right, it passes it as positional arg. I’ll have to add more tests to the example repo